src/state/block_tree.rs

```diff
@@ -0,0 +1,484 @@
+        let mut wb = BlockTreeWriteBatch::new();
+        wb.set_highest_view_entered(view)?;
+        self.write(wb);
+        Ok(())
+    }
+
+    pub fn set_highest_view_voted(&mut self, view: ViewNumber) -> Result<(), BlockTreeError> {
+        let mut wb = BlockTreeWriteBatch::new();
+        wb.set_highest_view_voted(view)?;
+        self.write(wb);
+        Ok(())
+    }
+
+    /* ↓↓↓ For deleting abandoned branches in insert_block ↓↓↓ */
+
+    /// Delete the "siblings" of the specified block, along with all of its associated data (e.g., pending
+    /// app state updates). Siblings
+    /// here refer to other blocks that share the same parent as the specified block.
+    ///
+    ///  # Precondition
+    /// Block is in its parents' (or the genesis) children list.
+    ///
+    /// # Error
+    /// Returns an error if the block is not in the block tree, or if the block's parent (or genesis) does not have a
+    /// children list.
+    pub fn delete_siblings(
+        &mut self,
+        wb: &mut BlockTreeWriteBatch<K::WriteBatch>,
+        block: &CryptoHash,
+    ) -> Result<(), BlockTreeError> {
+        let parent_or_genesis = self.block_justify(block)?.block;
+        let parents_or_genesis_children = self.children(&parent_or_genesis)?;
+        let siblings = parents_or_genesis_children
+            .iter()
+            .filter(|sib| *sib != block);
+        for sibling in siblings {
+            self.delete_branch(wb, sibling);
+        }
+
+        wb.set_children(&parent_or_genesis, &ChildrenList::new(vec![*block]))?;
+        Ok(())
+    }
+
+    /// Deletes all data of blocks in a branch starting from (and including) a given root block.
+    fn delete_branch(&mut self, wb: &mut BlockTreeWriteBatch<K::WriteBatch>, root: &CryptoHash) {
+        for block in self.blocks_in_branch(*root) {
+            wb.delete_children(&block);
+            wb.delete_pending_app_state_updates(&block);
+            wb.delete_block_validator_set_updates(&block);
+
+            if let Ok(Some(data_len)) = self.block_data_len(&block) {
+                wb.delete_block(&block, data_len)
+            }
+        }
+    }
+
+    /// Performs depth-first search to collect the hashes of all blocks in the branch rooted at `root` into
+    /// a single iterator.
+    pub fn blocks_in_branch(&self, root: CryptoHash) -> impl Iterator<Item = CryptoHash> {
+        let mut stack: Vec<CryptoHash> = vec![root];
+        let mut branch: Vec<CryptoHash> = vec![];
+
+        while let Some(block) = stack.pop() {
+            if let Ok(children) = self.children(&block) {
+                for child in children.iter() {
+                    stack.push(*child)
+                }
+            };
+            branch.push(block)
+        }
+        branch.into_iter()
+    }
+
+    /* ↓↓↓ Extra state getters for convenience ↓↓↓ */
+
+    pub fn contains(&self, block: &CryptoHash) -> bool {
+        self.block(block).is_ok_and(|block_opt| block_opt.is_some())
```

Comment:
This implementation of `contains` calls `block`, which means that checking whether the block tree contains a block entails reading the entire block from the KV Store, which is very inefficient.

I believe that v0.3 instead only gets one of the fields of block in order to check existence (iirc: `BLOCK_DATA_LEN`). This approach is more efficient.

Reply:
Will fix with https://github.com/parallelchain-io/hotstuff_rs/issues/44